{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Chapter 2\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Brute force algorithm for motif finding**  \n",
    "\n",
    "Given a collection of strings Dna and an integer d, a k-mer is a (k, d)-motif if it appears in every string from Dna with at most d mismatches. \n",
    "\n",
    "**Implanted motif problem**  \n",
    "Find all (k, d)-motifs in a collection of strings.  \n",
    "Input: A collection of strings Dna, and integers k and d.  \n",
    "Output: All (k,d)-motifs in Dna."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# from ch 1\n",
    "def HammingDist(patternA, patternB):\n",
    "    \"This function calculates the number of mismatches between two strings of equal length\"\n",
    "    # set number of mismatches to 0, convert input stirngs to list\n",
    "    number_mismatch = 0\n",
    "    assert len(patternA) == len(patternB), 'please input strings of equal length'\n",
    "    for i in range(len(patternA)):\n",
    "        if patternA[i] != patternB[i]:\n",
    "            number_mismatch += 1\n",
    "    return number_mismatch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def ApproxPattMatch(Pattern, Text, d):\n",
    "    '''\n",
    "    - This function returns the starting positions of a kmer\n",
    "    of interest within a DNA string. The kmer may have up to\n",
    "    d mismatches. Indices are 0 based.\n",
    "    - Pattern: motif of interest\n",
    "    - Text: DNA string \n",
    "    - d: tolerated number of mismatches\n",
    "    '''\n",
    "    # define range of Text to work in\n",
    "    overlap = len(Text) - len(Pattern) + 1\n",
    "    matches = []\n",
    "    # define variables to calculate Hamming Distance\n",
    "    # for every position in overlap:\n",
    "    for i in range(overlap):\n",
    "        # start is the position in overlap we loop over\n",
    "        start = i\n",
    "        # end is the position in overlap + the length of pattern\n",
    "        end = i + len(Pattern)\n",
    "        # string2 is the subset of text we are working with\n",
    "        subset_Text = Text[start:end]\n",
    "        \n",
    "        # calculate Hamming Distance between two strings\n",
    "        # If there are fewer mismatches than d, append\n",
    "        if HammingDist(Pattern, subset_Text) <= d:\n",
    "            matches.append(i)\n",
    "        \n",
    "    return matches"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "def MotifEnumerate(Dna, k, d):\n",
    "    '''\n",
    "    Dna: a list of DNA strings\n",
    "    k: k-mer motif\n",
    "    d: tolerated number of mismatches\n",
    "    '''\n",
    "    \n",
    "    Patterns = set()\n",
    "    # loop over the first DNA string in the list\n",
    "    # for Dna_string in Dna:\n",
    "    \n",
    "    # only need kmers that appear in the first DNA\n",
    "    # string, as each kmer must appear across all\n",
    "    # DNA strings?\n",
    "    \n",
    "    Dna_string1 = Dna[0]\n",
    "    overlap = len(Dna_string1) - k + 1\n",
    "    # get a kmer\n",
    "    for i in range(overlap):\n",
    "        start = i\n",
    "        end = i + k\n",
    "        kmer = Dna_string1[start:end]\n",
    "        for Dna_other in Dna:\n",
    "            print(kmer, \": \", Dna_other, ApproxPattMatch(Pattern = kmer, Text = Dna_other, d = d))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "DNA = ['TATCGA', 'ATGCA', 'ACGGT']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TAT :  TATCGA [0]\n",
      "TAT :  ATGCA []\n",
      "TAT :  ACGGT []\n",
      "ATC :  TATCGA [1]\n",
      "ATC :  ATGCA [0]\n",
      "ATC :  ACGGT []\n",
      "TCG :  TATCGA [2]\n",
      "TCG :  ATGCA []\n",
      "TCG :  ACGGT [0]\n",
      "CGA :  TATCGA [3]\n",
      "CGA :  ATGCA []\n",
      "CGA :  ACGGT [1]\n"
     ]
    }
   ],
   "source": [
    "MotifEnumerate(DNA, 3, 1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Luiz's solution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def motif_enumeration(text, k, distance):\n",
    "    patterns = set()\n",
    "    \n",
    "    first = text[0]\n",
    "    \n",
    "    for i in range(len(first) -k + 1):\n",
    "        pattern = first[i : i + k]\n",
    "        for pattern_p in neighbors(pattern, distance):\n",
    "            patterns.add(pattern_p)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 2b"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Median String Problem:\n",
    "\n",
    "Find Median String.  \n",
    "\n",
    "Input: A collection of stings Dna and an integer k.  \n",
    "Output: a k-mer Pattern minimizing d(Pattern, Dna) among all possible choices of k-mers."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Luiz's solution"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 2C\n",
    "\n",
    "Profile-most Probable k-mer Problem:  \n",
    "Find a Profile-most probable k-mer in a string.  \n",
    "\n",
    "Input: A string Text, an integer k, and a 4 x k matrix Profile  \n",
    "Output: A Porfile-most porbable k-mer in Text."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "def multiplyList(myList): \n",
    "    '''\n",
    "    Multiply elements one by one \n",
    "    '''\n",
    "    result = 1\n",
    "    for x in myList: \n",
    "         result = result * x  \n",
    "    return result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "def ProfileMostProbable(dna, k, profile):\n",
    "    '''\n",
    "    dna: a character string of DNA nucleotides\n",
    "    k: an integer specifying k size\n",
    "    mat: a matrix profile indicating nucleotide probabilities for each position in k-mer size k. \n",
    "    Given matrix Profile, evaluate the probability of every k-mer \n",
    "    in a string Text and find the Profile-most probable k-mer in \n",
    "    Text. If there are multiple, select the first.\n",
    "    '''\n",
    "    overlap = len(dna) - k + 1\n",
    "    # get a kmer\n",
    "    #kmer_prob_all = {}\n",
    "    kmer_prob_max = 0\n",
    "    max_kmer = \"\"\n",
    "    for i in range(overlap):\n",
    "        start = i\n",
    "        end = i + k\n",
    "        kmer = dna[start:end]\n",
    "        kmer_prob = 1 # must be 1! \n",
    "        for i, basepair in enumerate(kmer):\n",
    "            prob = profile[basepair][i]\n",
    "            kmer_prob *= prob\n",
    "        if  kmer_prob > kmer_prob_max:\n",
    "            kmer_prob_max = kmer_prob\n",
    "            max_kmer = kmer\n",
    "        #kmer_prob_all[kmer] = kmer_prob\n",
    "    print(max_kmer, kmer_prob_max)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "DNA = 'ACCTGTTTATTGCCTAAGTTCCGAACAAACCCAATATAGCCCGAGGGCCT'\n",
    "k = 5\n",
    "prof = {'A': [0.2, 0.2, 0.3, 0.2, 0.3],\n",
    "        'C': [0.4, 0.3, 0.1, 0.5, 0.1],\n",
    "        'G': [0.3, 0.3, 0.5, 0.2, 0.4],\n",
    "        'T': [0.1, 0.2, 0.1, 0.1, 0.2]\n",
    "       }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CCGAG 0.0048000000000000004\n"
     ]
    }
   ],
   "source": [
    "ProfileMostProbable(dna = DNA, k = k, profile = prof)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Tuple - immutable. Makes them faster for the python interpretter!\n",
    "def ProfileMostProbableTuple(dna, k, profile):\n",
    "    '''\n",
    "    dna: a character string of DNA nucleotides\n",
    "    k: an integer specifying k size\n",
    "    profile: a matrix profile indicating nucleotide probabilities for each position in k-mer size k. \n",
    "    Given matrix Profile, evaluate the probability of every k-mer \n",
    "    in a string Text and find the Profile-most probable k-mer in \n",
    "    Text. If there are multiple, select the first.\n",
    "    '''\n",
    "    overlap = len(dna) - k + 1\n",
    "    # get a kmer\n",
    "    kmer_prob_all = []\n",
    "    for i in range(overlap):\n",
    "        start = i\n",
    "        end = i + k\n",
    "        kmer = dna[start:end]\n",
    "        kmer_prof = []\n",
    "        for i, basepair in enumerate(kmer):\n",
    "            if basepair == 'A':\n",
    "                prob = profile.get('A')\n",
    "                prob = prob[i]\n",
    "                kmer_prof.append(prob)\n",
    "            elif basepair == \"T\":\n",
    "                prob = profile.get('T')\n",
    "                prob = prob[i]                \n",
    "                kmer_prof.append(prob)\n",
    "            elif basepair == \"C\":\n",
    "                prob = profile.get('C')\n",
    "                prob = prob[i] \n",
    "                kmer_prof.append(prob)\n",
    "            elif basepair == \"G\":\n",
    "                prob = profile.get('G')\n",
    "                prob = prob[i] \n",
    "                kmer_prof.append(prob)\n",
    "            else:\n",
    "                print(\"The string you input is not DNA, or is not formatted correctly. Please try again :)\")\n",
    "        kmer_prob = multiplyList(kmer_prof)\n",
    "        kmer_prob_all.append((kmer_prob, kmer)) \n",
    "    kmer_prob_all.sort() # sort method is in place; doesn't return new list\n",
    "    print(kmer_prob_all[-1][1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ProfileMostProbableTuple(dna = DNA, k = k, profile = prof)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 2D\n",
    "\n",
    "Greedy Motif Search\n",
    "\n",
    "Given: Integers k and t, followed by a collection of strings Dna.\n",
    "\n",
    "Return: A collection of strings BestMotifs resulting from running GreedyMotifSearch(Dna, k, t). If at any step you find more than one Profile-most probable k-mer in a given string, use the one occurring first."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "def profile(motifs, k):\n",
    "    matrix = []\n",
    "    # make positions for matrix\n",
    "    for i in range(4):\n",
    "        matrix.append([.0] * k)\n",
    "    # fill in matrix\n",
    "    n_motifs = len(motifs) # generate once bc constant\n",
    "    for i in range(k):\n",
    "        for motif in motifs:\n",
    "            motif_count = {'A': 0, 'C': 0, 'G': 0, 'T':0}\n",
    "            motif_count['A'] += motif[i].count('A')\n",
    "            motif_count['C'] += motif[i].count('C')\n",
    "            motif_count['G'] += motif[i].count('G')\n",
    "            motif_count['T'] += motif[i].count('T')\n",
    "        matrix[0][i] = motif_count['A'] / n_motifs\n",
    "        matrix[1][i] = motif_count['T'] / n_motifs\n",
    "        matrix[2][i] = motif_count['C'] / n_motifs\n",
    "        matrix[3][i] = motif_count['G'] / n_motifs\n",
    "        \n",
    "        \n",
    "    return matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "def GreedyMotifSearch(Dna, k, t):\n",
    "    '''\n",
    "    Dna: a collection of DNA strings\n",
    "    k: integer, k-mer length\n",
    "    t: integer, number of Dna strings in collection\n",
    "    '''\n",
    "    # motif matrix formed by first k-mers in each string from Dna\n",
    "    best_motifs = []\n",
    "    # don't need to iterate over list by position\n",
    "    # can just get each piece of dna list\n",
    "    #for seq in range(len(Dna)): \n",
    "    #    print(seq)\n",
    "    #    best_motifs.append(Dna[seq][0:k])\n",
    "    \n",
    "    # this is easier to read:\n",
    "    #for seq in Dna:\n",
    "    #    first_kmer = seq[0:k]\n",
    "    #    best_motifs.append(first_kmer)\n",
    "    \n",
    "    # But list comprehensions use more C, so they're faster\n",
    "    best_motifs = [seq[0:k] for seq in Dna]\n",
    "    \n",
    "    # improve the initial motif matrix by trying each k-mer in Dna1 as the first motifs\n",
    "    Dna1 = Dna[0]\n",
    "    overlap = len(Dna1) - k + 1\n",
    "    for start in range(overlap):\n",
    "        kmer = Dna1[start: start + k]\n",
    "        motif = [kmer]    \n",
    "        \n",
    "        for i in range(1, t): \n",
    "            # don't go through the first string\n",
    "            # first string is seed motif that we compare against\n",
    "            matrix = profile(motif, k = k)\n",
    "            print(*matrix, sep = '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "k = 3 \n",
    "t = 5\n",
    "DNA = [\"GGCGTTCAGGCA\",\n",
    "       \"AAGAATCAGTCA\",\n",
    "       \"CAAGGAGTTCGC\",\n",
    "       \"CACGTCAATCAC\",\n",
    "       \"CAATAATATTCG\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0.0, 0.0, 0.0]\n",
      "[0.0, 0.0, 0.0]\n",
      "[0.0, 0.0, 1.0]\n",
      "[1.0, 1.0, 0.0]\n",
      "[0.0, 0.0, 0.0]\n",
      "[0.0, 0.0, 0.0]\n",
      "[0.0, 0.0, 1.0]\n",
      "[1.0, 1.0, 0.0]\n",
      "[0.0, 0.0, 0.0]\n",
      "[0.0, 0.0, 0.0]\n",
      "[0.0, 0.0, 1.0]\n",
      "[1.0, 1.0, 0.0]\n",
      "[0.0, 0.0, 0.0]\n",
      "[0.0, 0.0, 0.0]\n",
      "[0.0, 0.0, 1.0]\n",
      "[1.0, 1.0, 0.0]\n",
      "[0.0, 0.0, 0.0]\n",
      "[0.0, 0.0, 0.0]\n",
      "[0.0, 1.0, 0.0]\n",
      "[1.0, 0.0, 1.0]\n",
      "[0.0, 0.0, 0.0]\n",
      "[0.0, 0.0, 0.0]\n",
      "[0.0, 1.0, 0.0]\n",
      "[1.0, 0.0, 1.0]\n",
      "[0.0, 0.0, 0.0]\n",
      "[0.0, 0.0, 0.0]\n",
      "[0.0, 1.0, 0.0]\n",
      "[1.0, 0.0, 1.0]\n",
      "[0.0, 0.0, 0.0]\n",
      "[0.0, 0.0, 0.0]\n",
      "[0.0, 1.0, 0.0]\n",
      "[1.0, 0.0, 1.0]\n",
      "[0.0, 0.0, 0.0]\n",
      "[0.0, 0.0, 1.0]\n",
      "[1.0, 0.0, 0.0]\n",
      "[0.0, 1.0, 0.0]\n",
      "[0.0, 0.0, 0.0]\n",
      "[0.0, 0.0, 1.0]\n",
      "[1.0, 0.0, 0.0]\n",
      "[0.0, 1.0, 0.0]\n",
      "[0.0, 0.0, 0.0]\n",
      "[0.0, 0.0, 1.0]\n",
      "[1.0, 0.0, 0.0]\n",
      "[0.0, 1.0, 0.0]\n",
      "[0.0, 0.0, 0.0]\n",
      "[0.0, 0.0, 1.0]\n",
      "[1.0, 0.0, 0.0]\n",
      "[0.0, 1.0, 0.0]\n",
      "[0.0, 0.0, 0.0]\n",
      "[0.0, 1.0, 1.0]\n",
      "[0.0, 0.0, 0.0]\n",
      "[1.0, 0.0, 0.0]\n",
      "[0.0, 0.0, 0.0]\n",
      "[0.0, 1.0, 1.0]\n",
      "[0.0, 0.0, 0.0]\n",
      "[1.0, 0.0, 0.0]\n",
      "[0.0, 0.0, 0.0]\n",
      "[0.0, 1.0, 1.0]\n",
      "[0.0, 0.0, 0.0]\n",
      "[1.0, 0.0, 0.0]\n",
      "[0.0, 0.0, 0.0]\n",
      "[0.0, 1.0, 1.0]\n",
      "[0.0, 0.0, 0.0]\n",
      "[1.0, 0.0, 0.0]\n",
      "[0.0, 0.0, 0.0]\n",
      "[1.0, 1.0, 0.0]\n",
      "[0.0, 0.0, 1.0]\n",
      "[0.0, 0.0, 0.0]\n",
      "[0.0, 0.0, 0.0]\n",
      "[1.0, 1.0, 0.0]\n",
      "[0.0, 0.0, 1.0]\n",
      "[0.0, 0.0, 0.0]\n",
      "[0.0, 0.0, 0.0]\n",
      "[1.0, 1.0, 0.0]\n",
      "[0.0, 0.0, 1.0]\n",
      "[0.0, 0.0, 0.0]\n",
      "[0.0, 0.0, 0.0]\n",
      "[1.0, 1.0, 0.0]\n",
      "[0.0, 0.0, 1.0]\n",
      "[0.0, 0.0, 0.0]\n",
      "[0.0, 0.0, 1.0]\n",
      "[1.0, 0.0, 0.0]\n",
      "[0.0, 1.0, 0.0]\n",
      "[0.0, 0.0, 0.0]\n",
      "[0.0, 0.0, 1.0]\n",
      "[1.0, 0.0, 0.0]\n",
      "[0.0, 1.0, 0.0]\n",
      "[0.0, 0.0, 0.0]\n",
      "[0.0, 0.0, 1.0]\n",
      "[1.0, 0.0, 0.0]\n",
      "[0.0, 1.0, 0.0]\n",
      "[0.0, 0.0, 0.0]\n",
      "[0.0, 0.0, 1.0]\n",
      "[1.0, 0.0, 0.0]\n",
      "[0.0, 1.0, 0.0]\n",
      "[0.0, 0.0, 0.0]\n",
      "[0.0, 1.0, 0.0]\n",
      "[0.0, 0.0, 0.0]\n",
      "[1.0, 0.0, 0.0]\n",
      "[0.0, 0.0, 1.0]\n",
      "[0.0, 1.0, 0.0]\n",
      "[0.0, 0.0, 0.0]\n",
      "[1.0, 0.0, 0.0]\n",
      "[0.0, 0.0, 1.0]\n",
      "[0.0, 1.0, 0.0]\n",
      "[0.0, 0.0, 0.0]\n",
      "[1.0, 0.0, 0.0]\n",
      "[0.0, 0.0, 1.0]\n",
      "[0.0, 1.0, 0.0]\n",
      "[0.0, 0.0, 0.0]\n",
      "[1.0, 0.0, 0.0]\n",
      "[0.0, 0.0, 1.0]\n",
      "[1.0, 0.0, 0.0]\n",
      "[0.0, 0.0, 0.0]\n",
      "[0.0, 0.0, 0.0]\n",
      "[0.0, 1.0, 1.0]\n",
      "[1.0, 0.0, 0.0]\n",
      "[0.0, 0.0, 0.0]\n",
      "[0.0, 0.0, 0.0]\n",
      "[0.0, 1.0, 1.0]\n",
      "[1.0, 0.0, 0.0]\n",
      "[0.0, 0.0, 0.0]\n",
      "[0.0, 0.0, 0.0]\n",
      "[0.0, 1.0, 1.0]\n",
      "[1.0, 0.0, 0.0]\n",
      "[0.0, 0.0, 0.0]\n",
      "[0.0, 0.0, 0.0]\n",
      "[0.0, 1.0, 1.0]\n",
      "[0.0, 0.0, 0.0]\n",
      "[0.0, 0.0, 0.0]\n",
      "[0.0, 0.0, 1.0]\n",
      "[1.0, 1.0, 0.0]\n",
      "[0.0, 0.0, 0.0]\n",
      "[0.0, 0.0, 0.0]\n",
      "[0.0, 0.0, 1.0]\n",
      "[1.0, 1.0, 0.0]\n",
      "[0.0, 0.0, 0.0]\n",
      "[0.0, 0.0, 0.0]\n",
      "[0.0, 0.0, 1.0]\n",
      "[1.0, 1.0, 0.0]\n",
      "[0.0, 0.0, 0.0]\n",
      "[0.0, 0.0, 0.0]\n",
      "[0.0, 0.0, 1.0]\n",
      "[1.0, 1.0, 0.0]\n",
      "[0.0, 0.0, 1.0]\n",
      "[0.0, 0.0, 0.0]\n",
      "[0.0, 1.0, 0.0]\n",
      "[1.0, 0.0, 0.0]\n",
      "[0.0, 0.0, 1.0]\n",
      "[0.0, 0.0, 0.0]\n",
      "[0.0, 1.0, 0.0]\n",
      "[1.0, 0.0, 0.0]\n",
      "[0.0, 0.0, 1.0]\n",
      "[0.0, 0.0, 0.0]\n",
      "[0.0, 1.0, 0.0]\n",
      "[1.0, 0.0, 0.0]\n",
      "[0.0, 0.0, 1.0]\n",
      "[0.0, 0.0, 0.0]\n",
      "[0.0, 1.0, 0.0]\n",
      "[1.0, 0.0, 0.0]\n"
     ]
    }
   ],
   "source": [
    "GreedyMotifSearch(Dna = DNA, k = k, t = t)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
